updates_and_signals/atomic_message_handlers/workflow.py

```diff
@@ -0,0 +1,198 @@
+
+
+@dataclass(kw_only=True)
+class ClusterManagerInput:
+    state: Optional[ClusterManagerState] = None
+    test_continue_as_new: bool = False
+
+
+@dataclass
+class ClusterManagerResult:
+    max_assigned_nodes: int
+    num_assigned_nodes: int
+
+
+# ClusterManagerWorkflow keeps track of the allocations of a cluster of nodes.
+# Via signals, the cluster can be started and shutdown.
+# Via updates, clients can also assign jobs to nodes and delete jobs.
+# These updates must run atomically.
+@workflow.defn
+class ClusterManagerWorkflow:
+    def __init__(self) -> None:
+        self.state = ClusterManagerState()
+        # Protects workflow state from interleaved access
+        self.nodes_lock = asyncio.Lock()
+
+    @workflow.signal
+    async def start_cluster(self):
+        self.state.cluster_started = True
+        self.state.nodes = {str(k): None for k in range(25)}
+        workflow.logger.info("Cluster started")
+
+    @workflow.signal
+    async def shutdown_cluster(self):
+        await workflow.wait_condition(lambda: self.state.cluster_started)
+        self.state.cluster_shutdown = True
+        workflow.logger.info("Cluster shut down")
+
+    @workflow.update
+    async def allocate_n_nodes_to_job(
+        self,
+        task_name: str,
+        num_nodes: int,
+    ) -> List[str]:
+        await workflow.wait_condition(lambda: self.state.cluster_started)
+        if self.state.cluster_shutdown:
+            raise RuntimeError(
+                "Cannot allocate nodes to a job: Cluster is already shut down"
+            )
+
+        async with self.nodes_lock:
+            unassigned_nodes = [k for k, v in self.state.nodes.items() if v is None]
+            if len(unassigned_nodes) < num_nodes:
+                raise ValueError(
+                    f"Cannot allocate {num_nodes} nodes; have only {len(unassigned_nodes)} available"
+                )
+            assigned_nodes = unassigned_nodes[:num_nodes]
+            # This await would be dangerous without nodes_lock because it yields control and allows interleaving.
+            await self._allocate_nodes_to_job(assigned_nodes, task_name)
+            self.state.max_assigned_nodes = max(
+                self.state.max_assigned_nodes,
+                len([k for k, v in self.state.nodes.items() if v is not None]),
+            )
+            return assigned_nodes
+
+    async def _allocate_nodes_to_job(self, assigned_nodes: List[str], task_name: str):
+        await workflow.execute_activity(
+            allocate_nodes_to_job,
+            AllocateNodesToJobInput(nodes=assigned_nodes, task_name=task_name),
+            start_to_close_timeout=timedelta(seconds=10),
+        )
+        for node in assigned_nodes:
+            self.state.nodes[node] = task_name
+
+    @workflow.update
+    async def delete_job(self, task_name: str) -> str:
+        await workflow.wait_condition(lambda: self.state.cluster_started)
+        assert not self.state.cluster_shutdown
```

Comment:
This should probably match the error from allocate (see comment there, this will fail task by default, may prefer `ApplicationError`)